### ISSUE TITLE
Making allowances for an incremental build process

### ISSUE THREAD

**Issue by Seanny123** [nengo/nengo#869]
As discussed in #867 adding some sort of ability for an incremental build process would be useful.

@jgosmann wants it for SpaOpt
@arvoelke used it for his touch classification paper
@hunse maybe mentioned it for deep belief networks?

I thought I would open a new thread for discussion so this point doesn't get lost to history.

**Comment by arvoelke** [comment]
Thanks. :smile: Disclaimer: it's not at all clear to me how such a process should look from the user's perspective and how this might affect the bigger picture (e.g. nengo_gui, other backends). I have at least two ideas in mind that are very different. I'm happy to brainstorm a bit if we want to take this seriously for the future.

**Comment by Seanny123** [comment]
This seems like a pretty big topic with a lot of things to consider. Maybe we should postpone discussion until the next Nengo meeting (tomorrow)?

**Comment by hunse** [comment]
I would postpone for a while. Getting 2.1 out is more of a priority. If people want to meet in small groups to discuss this, that's fine, and actually I think that kind of brainstorming should happen before we bring this to the big group.

**Comment by Seanny123** [comment]
Sounds good to me.

**Comment by arvoelke** [comment]
Agreed. This is a tricky topic that will take some time to think over.

**Comment by tcstewar** [comment]
Another use case for this: Parisien transforms   #921

**Comment by tcstewar** [comment]
As a quick update, I finally got around to implementing the Parisien transform and so I at least know now what information is needed from the build process for this use case.

The code's in #921 but the basic information needed is:
 - the pre-synaptic decoders
 - the post-synaptic encoders
 - the activity matrices and eval_points for the pre-synaptic neurons
 - the synapse used by the Connection
 - the number of neurons in the pre-synaptic ensemble

Given all of that, it creates three new Connections and a new Ensemble (the interneurons)

Anyway, not sure how all that fits into people's thinking, but I thought I'd post that list here for posterity.

**Comment by jgosmann** [comment]
I finally got spaopt based on the deferrables into a working state (#941) and I still favour this approach over a helper function that modifies the model after constructing it.

Such a function would need different code for `Bind`, `Compare`, and `State` because, even though the radius is derived with the same optimization method, the final radius is adjusted by a different factor in each case. That factor can even depend on the exact implementation of the underlying networks (and this implementation could change in the future). Thus, the helper function would need to know a great deal about the network internals which doesn't seem like good code structure to me and it would not be obvious at all that changes to one of those networks require changes to spaopt.

It also seems difficult to support user-written modules (that construct their own low-level representation ensembles and do not rely on aforementioned modules) with the helper function approach because that function needs to know for each module how to determine the radius, but it cannot know this for use modules. With the deferrables approach the user-written module just has to get the base radius from whatever was passed in or configured as `radius_method` and can then apply any adjustments it needs.

Overall, I would assume that a helper function for spaopt would not be any less code or complexity (though complexity of a different kind).


If we don't want the deferrables system, there could be one more possibility for a spaopt implementation, where one would have to create an instance of a `SpaOpt` class with `Simulator` class passed in and assign it to the `radius_method` config parameter. However, this would require to know the simulator being used upfront which might be a problem (for example if you run the model in nengo_gui and change the backend in the GUI you also have to edit it in the code). Also, this would require to calculate the radius without knowing the final ensemble parameters which could be a problem.

This solution could not be reused for similar cases (e.g. neuron number to meet accuracy criterion) without adding special code to some underlying Nengo objects. (For spaopt this is not relevant, because the code has to be added in either case because of other reasons.) With deferrables such other cases could probably implemented without any other special case code.

That being said, I feel that a helper function might be the right approach for the Parisien transform which actually changes the model structure.

**Comment by jgosmann** [comment]
I started implementing the method for determining the neuron number given the desired accuracy. I'm not finished yet, but it seems that it requires exactly the same things as spaopt. I'm also thinking about the possibility of an alternate method to determine this neuron number where the same thing applies, but additionally it would require incrementally adding neurons to an ensemble. In that case it is not required to recompute the complete inverse matrix each time, but a less costly update is possible. It might be possible to even do this without special support in the builder by implementing a special solver that supports this incremental process. But potentially this will need access to the seeds (which I believe I can already get with the current implementation of the builder and deferrables).

**Comment by jgosmann** [comment]
As promised I implemented the method (as proof of concept) for determining the neuron numbers: #1062.

That gives now two examples for using deferrables and they both need the same information. [I make an argument why this is the right set of arguments](https://github.com/nengo/nengo/pull/995#issuecomment-218588383) in PR #995.

**Comment by drasmuss** [comment]
We plan to add something like this as part of the builder refactoring (which is tracked in other issues).
